core/pkg/artifacts/saver.go

```diff
@@ -144,99 +146,157 @@ func (as *ArtifactSaver) uploadFiles(artifactID string, manifest *Manifest, mani
-				}
-				numInProgress++
-				task := &filetransfer.Task{
-					FileKind: filetransfer.RunFileKindArtifact,
-					Type:     filetransfer.UploadTask,
-					Path:     *entry.LocalPath,
-					Url:      *edge.Node.UploadUrl,
-					Headers:  edge.Node.UploadHeaders,
-				}
-				task.SetCompletionCallback(
-					func(t *filetransfer.Task) {
-						taskResultsChan <- TaskResult{t, name}
-					},
-				)
-				as.FileTransferManager.AddTask(task)
-			}
 		}
-		// Wait for filetransfer to catch up. If there's nothing more to schedule, wait for all in progress tasks.
-		for numInProgress > maxBacklog || (len(fileSpecsBatch) == 0 && numInProgress > 0) {
-			numInProgress--
-			result := <-taskResultsChan
-			if result.Task.Err != nil {
-				// We want to retry when the signed URL expires. However, distinguishing that error from others is not
-				// trivial. As a heuristic, we retry if the request failed more than an hour after we fetched the URL.
-				if time.Since(nameToScheduledTime[result.Name]) < 1*time.Hour {
-					return result.Task.Err
-				}
-				delete(nameToScheduledTime, result.Name) // retry
+		if len(fileSpecs) > 0 {
+			slog.Warn("some files failed to upload, retrying", "count", len(fileSpecs))
+		}
+	}
+	return nil
+}
+
+func (as *ArtifactSaver) processFiles(
+	manifest *Manifest, fileSpecs map[string]gql.CreateArtifactFileSpecInput,
+) (map[string]gql.CreateArtifactFileSpecInput, error) {
+
+	// Channels to get responses from the batch url retrievers.
+	readyChan := make(chan serverFileResponse, as.batchSize)
+	errorChan := make(chan error)
+
+	doneChan := make(chan *filetransfer.Task, as.batchSize)
+	mustRetry := map[string]gql.CreateArtifactFileSpecInput{}
+
+	numTotal := len(fileSpecs)
+	numActive := 0
+	numDone := 0
+
+	fileSpecBatches := batchify(fileSpecs, as.batchSize)
+
+	for numDone+len(mustRetry) < numTotal {
+		// If we have less than a batch worth of requests open, start a new batch.
+		if numActive <= as.batchSize && len(fileSpecBatches) > 0 {
+			batch := fileSpecBatches[0]
+			numActive += len(batch)
+			go as.batchFileDataRetriever(batch, readyChan, errorChan)
+			fileSpecBatches = fileSpecBatches[1:]
+		}
+		select {
+		// Start any uploads that are ready.
+		case fileInfo := <-readyChan:
+			entry := manifest.Contents[fileInfo.Name]
+			entry.BirthArtifactID = &fileInfo.BirthArtifactID
+			manifest.Contents[fileInfo.Name] = entry
+			cacheEntry(as.FileCache, entry)
+			if fileInfo.UploadUrl == nil {
+				// The server already has this file.
+				numActive--
+				numDone++
 				continue
 			}
-			numDone++
-			entry := manifest.Contents[result.Name]
-			if !entry.SkipCache {
-				digest := entry.Digest
-				go func() {
-					err := as.FileCache.AddFileAndCheckDigest(result.Task.Path, digest)
-					if err != nil {
-						slog.Error("error adding file to cache", "err", err)
-					}
-				}()
+			task := newUploadTask(fileInfo, *entry.LocalPath)
+			task.SetCompletionCallback(func(t *filetransfer.Task) { doneChan <- t })
```

Comment:
hmm, given how simple the logic is in the `doneChan` handling, can the callback just do it?

Reply:
We can, but because we're adding to a shared data structure we either need to put stuff into a channel anyway (so we wouldn't really simplify anything) or we need to protect it with a mutex. I haven't tried the mutex version because I was trying to avoid needing extra synchronization primitives, but it probably wouldn't be the bottleneck and shouldn't slow it down too much. Do you think you would like that form better?

Reply:
I would consider doing this so we dont have 2 workload channels to multi-plex over

Reply:
I played around with this a little and IMHO it muddies up the code a bit more than it helps, mostly around loop termination and channel closure. Going go punt on this for now.